xbmc/FileItemList.cpp

```diff
@@ -718,279 +742,164 @@ void CFileItemList::Stack(bool stackFiles /* = true */)
         {
-          std::string dvdPath = VIDEO::UTILS::GetOpticalMediaPath(*item);
-
-          if (!dvdPath.empty())
+          if (regExp.RegFind(file) != -1)
           {
-            // NOTE: should this be done for the CD# folders too?
-            item->m_bIsFolder = false;
-            item->SetPath(dvdPath);
-            item->SetLabel2("");
-            item->SetLabelPreformatted(true);
-            m_sortDescription.sortBy = SortByNone; /* sorting is now broken */
+            // Get components of file name
+            stackCandidates.emplace_back(FILE_CANDIDATE, regExp.GetMatch(1) /* Title */,
+                                         regExp.GetMatch(2) /* Volume */,
+                                         static_cast<int>(&item - &m_items[0]), item->m_dwSize);
+            break;
           }
         }
       }
     }
-  }
-}
-
-void CFileItemList::StackFiles()
-{
-  // Precompile our REs
-  VECCREGEXP stackRegExps;
-  CRegExp tmpRegExp(true, CRegExp::autoUtf8);
-  const std::vector<std::string>& strStackRegExps =
-      CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_videoStackRegExps;
-  std::vector<std::string>::const_iterator strRegExp = strStackRegExps.begin();
-  while (strRegExp != strStackRegExps.end())
-  {
-    if (tmpRegExp.RegComp(*strRegExp))
-    {
-      if (tmpRegExp.GetCaptureTotal() == 4)
-        stackRegExps.push_back(tmpRegExp);
-      else
-        CLog::Log(LOGERROR, "Invalid video stack RE ({}). Must have 4 captures.", *strRegExp);
-    }
-    ++strRegExp;
-  }
-
-  // now stack the files, some of which may be from the previous stack iteration
-  int i = 0;
-  while (i < Size())
-  {
-    CFileItemPtr item1 = Get(i);
-
-    // skip folders, nfo files, playlists
-    if (item1->m_bIsFolder || item1->IsParentFolder() || item1->IsNFO() ||
-        PLAYLIST::IsPlayList(*item1))
-    {
-      // increment index
-      i++;
-      continue;
-    }
 
-    int64_t size = 0;
-    size_t offset = 0;
-    std::string stackName;
-    std::string file1;
-    std::string filePath;
-    std::vector<int> stack;
-    VECCREGEXP::iterator expr = stackRegExps.begin();
+    // Sort stack candidates
+    std::sort(stackCandidates.begin(), stackCandidates.end(),
+              [](const auto& s, const auto& t)
+              {
+                if (std::get<0>(s) == std::get<0>(t)) // type (file/folder)
+                {
+                  if (std::get<1>(s) == std::get<1>(t)) // title
+                    return std::get<2>(s) < std::get<2>(t); // volume
+                  else
+                    return std::get<1>(s) < std::get<1>(t);
+                }
+                else
+                  return std::get<0>(s) < std::get<0>(t);
+              });
 
-    URIUtils::Split(item1->GetPath(), filePath, file1);
-    if (URIUtils::HasEncodedFilename(CURL(filePath)))
-      file1 = CURL::Decode(file1);
+    // Count stack candidates
+    std::map<std::pair<int, std::string>, int> countCandidates;
```

Comment:
A comment explaining what exactly the map contains would really help the code reader.